README

         M U M B L E
		
	  Audio Conference Tool Over NDN
	  redmine.irl.cs.ucla.edu/projects/ndn-act/repository

Mumble is a voicechat program for gamers written on top of Qt and Speex.

There are two modules in Mumble; the client (mumble) and the server
(murmur). The client works on Win32/64, Linux and Mac OS X, while the
server should work on anything Qt can be installed on.

Note that when we say Win32, we mean Windows XP or newer.

Running Mumble
==============

On Windows, after installation, you should have a new Mumble folder in your
Start Menu, from which you can start Mumble.

On Mac OS X, to install Mumble, drag the application from the downloaded
disk image into your /Applications folder.

Once Mumble is launched, you need a server to connect to. Either create your
own or join a friend's.

Running Murmur on Unix-like systems
===================================

Murmur should be run from the command line, so start a shell (command prompt)
and go to wherever you installed Mumble. Run murmur as

murmurd [-supw <password>] [-ini <inifile>] [-fg] [v]

-supw   Set new password for the user SuperUser, which is hardcoded to
        bypass ACLs. Keep this password safe. Until you set a password,
        the SuperUser is disabled. If you use this option, murmur will
        set the password in the database and then exit.

-ini    Use a inifile other than murmur.ini, use this to run several isntances
        of murmur from the same directory. Make sure each instance is using
        a separate database.

-fg     Run in the foreground, logging to standard output.

-v      More verbose logging.

Running Murmur on Mac OS X
==========================

On Mac OS X, Murmur is available inside your Mumble application bundle. If you
copied the Mumble program into your Applications directory, you should be able
to run it by executing:

 /Applications/Mumble.app/Contents/MacOS/murmurd -v -fg

in a Terminal. For more information, please see the 'Running Murmur on Unix-
like sytems' above.

To easily override the default Murmur configuration, a murmur.ini file can be placed in
your user's 'Library/Preferences/Mumble/' directory.

Example configuration files and other scripts can be found inside the Murmur
directory of the Mumble installation disk image.

Running Murmur on Win32
=======================

Doubleclick the Murmur icon to start murmur. There will be a small icon on your
taskbar from which you can view the log.

To set the superuser password, run murmur with the parameters '-supw <password>'.

Bandwidth usage
===============

Mumble will use 10-40 kbit/s outgoing, and the same incoming for each user.
So if there are 10 other users on the server with you, your incoming
bandwidth requirement will be 100-400 kbit/s if they all talk at the same time.


CHANGES
commit c52381a082e7ddcb5cbd34ba20f17081ecb7d1f2
Author: Zhenkai Zhu <zhenkai@cs.ucla.edu>
Date:   Thu Aug 25 22:22:43 2011 -0700

    do not try to create actd_keystore

commit 1487c41ca5563b954ef246dbf44ae09660761952
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Thu Aug 25 12:15:22 2011 -0700

    default to be "continuous" instead of VAD

commit 1b543c2b9c66d1ad1939711d5139e9a9d75ef880
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Thu Aug 25 11:43:37 2011 -0700

    server never ban user

commit 5e52c705ed080080c3e606150bfa8aca3515bdf3
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Thu Aug 25 10:29:57 2011 -0700

    no Bonjour

commit 8147149500ae2ba0add16acc7f9e29a2e72a0230
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Thu Aug 25 10:24:58 2011 -0700

    add localhost as server if no favorite exists

commit 08db9436fc968ef8e24ec07212c5ab71efe23138
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Thu Aug 25 09:48:14 2011 -0700

    scripts/ini.* changed

commit 1567f175d2e76b0743f7168e93d50cd8ca13d491
Author: Zhenkai Zhu <zhenkai@cs.ucla.edu>
Date:   Tue Aug 23 18:19:04 2011 -0700

    server bind to 127.0.0.1

commit 14f99792aed4fe780e38e27a7ac89c36cc1ab50a
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 22 14:37:03 2011 -0700

    pthread lock instead of trylock in media

commit ccb7cab2e6b3489fad15102466fbae7a846bc26d
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 22 12:05:10 2011 -0700

    no separate thread for ccn_run in media

commit 01d41b3419f774e99f56f80015799652d62a95b4
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 22 11:23:44 2011 -0700

     hard code quality and audio frames

commit 6d1c4ced76060d6047e5308df3cc528aa6cd16bb
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 22 11:21:34 2011 -0700

    no bandwidth limit

commit af59123e43ba474caf86b4feb37b3824af591049
Author: Zhenkai Zhu <zhenkai@cs.ucla.edu>
Date:   Thu Aug 18 22:43:21 2011 -0400

    update changes

commit ff3da370896716fc5dc3fc6cbd9e0f19a9245757
Author: Zhenkai Zhu <zhenkai@cs.ucla.edu>
Date:   Thu Aug 18 22:19:58 2011 -0400

    per packet len = 40

commit 48103b80402615304beccc6607f442001751439b
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Wed Aug 10 17:36:04 2011 -0700

    per packet len = 20

commit d6c4a8ae427832ed9c7c55ab3f2ec973fd86a02c
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Wed Aug 10 15:36:00 2011 -0700

    Revert "always 40ms per second"
    
    This reverts commit 18a839baeee3127136e4f6112c39f99252e8a625.

commit 51599febd8d001248b5bf742766da480bbbf83ca
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Wed Aug 10 15:35:49 2011 -0700

    Revert "frames = 2; maybe this is 40 ms"
    
    This reverts commit 897289461e469a57619eb1cc5af3547fd44a6a09.

commit f45ba105523571785f386a5c3ff52f97201d6289
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Wed Aug 10 13:15:49 2011 -0700

    usleep(200) in while loop when trylock

commit 897289461e469a57619eb1cc5af3547fd44a6a09
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 8 17:39:37 2011 -0700

    frames = 2; maybe this is 40 ms

commit 027d118d3e818afc0360dff85b7a9c59eb0001ed
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 8 17:37:45 2011 -0700

    check ru exist before check isStale

commit 18a839baeee3127136e4f6112c39f99252e8a625
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 8 17:27:38 2011 -0700

    always 40ms per second

commit 5522805113c77a544b58039c0561309d03f7e9eb
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 8 13:57:21 2011 -0700

    lock of actd

commit 3434683712e8f4b9553078a9164f85f6e7364c94
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 8 11:51:22 2011 -0700

    check timestamp is not null

commit 2a4c115a397a443a76215ac94e2f5810bfe33cbf
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 8 11:24:57 2011 -0700

    pop a qmsgbox and complain if ccn_connect failed

commit 68160494ae95433906a8e68eb67aa05f2a4ccb57
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Mon Aug 8 10:47:07 2011 -0700

    mac packaging scripts changed; changelog added

commit 5beaff36111bc44ed15fd9fe98ac9e9548580a90
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Fri Aug 5 18:41:23 2011 -0700

    correct poll
    needs to call ccn_run before poll so that the interest filter could be set up; otherwise ccnd will not try to send us intereset and poll will always return zero.

commit 35e567dcae5007ffe54cdf7a49c3886a768ee949
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Fri Aug 5 17:57:14 2011 -0700

    voice content freshness = 2
    if freshness = 4, the short interests would possibly be satisfied by the cached content, and trigger an unnecessiry init pipe and also makes tick() work uncessarily

commit affe0707274f8ec48b814a146c2ada077aa3717a
Author: Zhenkai Zhu <zzhu@parc.com>
Date:   Fri Aug 5 17:40:25 2011 -0700

    two callback functions for init and pipe ints

commit e33fde47c77ff1a1ad6aa18f3149ebfdf964a141
Author: zzhu <zzhu@ndn01.parc.xerox.com>
Date:   Fri Aug 5 11:35:08 2011 -0700

    disable daemonization of ndn-murmurd

commit d8388f9f46ece9559c113971aa3b23185d405e9d
Author: zzhu <zzhu@ndn01.parc.xerox.com>
Date:   Fri Aug 5 10:39:41 2011 -0700

    fallback to short interests when timeout works

commit c24dea8e04511240536033418a2b56d7727706ec
Author: zzhu <zzhu@ndn01.parc.xerox.com>
Date:   Fri Aug 5 10:19:58 2011 -0700

    calloc

commit d604f7a920f78644f1353256248276e633c2fb84
Author: zzhu <zzhu@ndn01.parc.xerox.com>
Date:   Fri Aug 5 10:06:11 2011 -0700

    intdata

commit 729a7023c9168559f432fe5ddb16b6d738c68914
Author: zzhu <zzhu@ndn01.parc.xerox.com>
Date:   Thu Aug 4 17:22:44 2011 -0700

    separate closure for short and pipe interests

commit 33f4087d6aeecf1aa93da0b57695a0c06bd08544
Author: zzhu <zzhu@ndn01.parc.xerox.com>
Date:   Thu Aug 4 16:55:12 2011 -0700

    only reexpress probe interest

commit 041995119fd29a0dfcbabedc35dd0206b57731fb
Author: zzhu <zzhu@ndn01.parc.xerox.com>
Date:   Thu Aug 4 16:21:45 2011 -0700

    check userBuf not null

commit 48f6c73929c3e017d237c6aba3e26c26ee78b5d7

mainwindow.cpp
#include <QtGui>
#include <QInputDialog>
#include <QFile>
#include <QDir>
#include <QTextStream>

#include "mainwindow.h"
#include "debugbox.h"


MainWindow::MainWindow(char *argv[], QWidget *parent) 
	:QDialog(parent)
{
	readSettings();

	binaryPath = argv[0];
	QString temp = binaryPath.left(binaryPath.lastIndexOf("/"));
	binaryPath = temp;
	debug("binary path is " + temp);
	pubConfList = new QTreeWidget;
	pubConfList->setEnabled(false);
	pubConfList->setRootIsDecorated(false);
	pubConfList->setHeaderLabels(QStringList() << tr("")<<tr("")<<tr("")<<tr("Conference Name")<<tr("Organizer")<<tr("Contact"));
	
	pubConfList->setColumnWidth(0, 25);
	pubConfList->setColumnWidth(1, 25);
	pubConfList->setColumnWidth(2, 25);
	pubConfList->setColumnWidth(3, 200);
	pubConfList->setColumnWidth(4, 150);


	//myConfList = new QTreeWidget;
	//myConfList->setEnabled(false);
	//myConfList->setRootIsDecorated(false);
	//myConfList->setHeaderLabels(QStringList() << tr("Audio")<<tr("Video")<<tr("Conference Name")<<tr("Organizer"));

	confDesc = new QTextEdit;
	confDesc->setReadOnly(true);
	confDesc->setEnabled(false);

	pubConfLabel = new QLabel(tr("Public Conferences"));
	//myConfLabel = new QLabel(tr("My Conferences"));
	confDescLabel = new QLabel(tr("Conference Descriptions"));
	QString warning = "Current prefix for voice data is: " + prefix ;
	currentPrefLabel = new QLabel(warning);

	newButton = new QPushButton(tr("New"));
	prefButton = new QPushButton(tr("Preferences"));
	aboutButton = new QPushButton(tr("About"));
	quitButton = new QPushButton(tr("Quit"));

	joinButton = new QPushButton(tr("Join"));
	joinButton->setEnabled(false);
	//editButton = new QPushButton(tr("Edit"));
	//editButton->setEnabled(false);
	dismissButton = new QPushButton(tr("Dismiss"));
	dismissButton->setEnabled(false);

	listPrivate = new QCheckBox(tr("Also List Private Conferences"));
	connect(listPrivate, SIGNAL(stateChanged(int)), this, SLOT(listPrivateConferences()));
	// TODO: should enable this in near future
	listPrivate->setEnabled(false);
	
	QHBoxLayout *topLayout = new QHBoxLayout;
	topLayout->addWidget(newButton);
	topLayout->addWidget(prefButton);
	topLayout->addWidget(aboutButton);
	topLayout->addWidget(quitButton);
	

	footButtonBox = new QDialogButtonBox;
	footButtonBox->addButton(joinButton, QDialogButtonBox::ActionRole);
	//footButtonBox->addButton(editButton, QDialogButtonBox::ActionRole);
	footButtonBox->addButton(dismissButton, QDialogButtonBox::ActionRole);

	QHBoxLayout *midLayout = new QHBoxLayout;
	midLayout->addWidget(pubConfLabel);
	midLayout->addWidget(listPrivate);

	QVBoxLayout *mainLayout = new QVBoxLayout;
	mainLayout->addLayout(topLayout);
	mainLayout->addLayout(midLayout);
	mainLayout->addWidget(pubConfList);
	mainLayout->addWidget(confDescLabel);
	mainLayout->addWidget(confDesc);
	mainLayout->addWidget(currentPrefLabel);
	mainLayout->addWidget(footButtonBox);
	setLayout(mainLayout);

	setWindowTitle(tr("Conference Management Tool"));

	connect(newButton, SIGNAL(clicked()), this, SLOT(newConference()));
	connect(prefButton, SIGNAL(clicked()), this, SLOT(changePref()));
	connect(aboutButton, SIGNAL(clicked()), this, SLOT(about()));
	connect(quitButton, SIGNAL(clicked()), this, SLOT(close()));
	connect(joinButton, SIGNAL(clicked()), this, SLOT(joinConference()));
	//connect(editButton, SIGNAL(clicked()), this, SLOT(editConference()));
	connect(dismissButton, SIGNAL(clicked()), this, SLOT(dismissConference()));
	connect(pubConfList, SIGNAL(currentItemChanged(QTreeWidgetItem *, QTreeWidgetItem *)), this, SLOT(processItem()));
	

	
	sd = new SessionEnum(prefix);
	connect(sd, SIGNAL(expired(QString, QString)), this, SLOT(removeConferenceFromList(QString, QString)));
	connect(sd, SIGNAL(add(Announcement *)), this, SLOT(addConferenceToList(Announcement *)));
	
}

QSize MainWindow::sizeHint() const
{
	return QSize(600, 600);
}


void MainWindow::processItem(){
	QTreeWidgetItem *current = pubConfList->currentItem();
	if (!current) {
		pubConfList->setEnabled(false);
		confDesc->clear();
		confDesc->setEnabled(false);
		joinButton->setEnabled(false);
		//editButton->setEnabled(false);
		dismissButton->setEnabled(false);
		return;
	}
	
	Announcement *a = itemToAnnouncement[current];
	if (a->getOwner()) {
		joinButton->setEnabled(true);
		//editButton->setEnabled(true);
		dismissButton->setEnabled(true);

	} else {
		joinButton->setEnabled(true);
		//editButton->setEnabled(false);
		dismissButton->setEnabled(false);
	}
//	confDesc->setPlainText(a->getDesc());
	QString desc = current->data(3, Qt::UserRole).toString();
	confDesc->setPlainText(desc);

}



void MainWindow::readSettings() {
	QSettings settings("UCLA_IRL", "ACTD");
	prefix = settings.value("prefix", QString("")).toString();

	if (prefix == "") {
		changePref();
	}

	audioPath = settings.value("audioPath", QString("")).toString();

}

void MainWindow::writeSettings() {
	QSettings settings("UCLA_IRL", "ACTD");
	settings.setValue("prefix", prefix);
	settings.setValue("audioPath", audioPath);
}

void MainWindow::about() {
	QMessageBox::about(this, tr("About actd"), QString("%1\n%2").arg(tr("This is an application similar to sd of the MBone.")).arg(tr("Conference management is handled by this application.")));

}


void MainWindow::changePref() {
	bool ok;

	QString text = QInputDialog::getText(this, tr("Get Name Prefix for Conference"), tr("Input the name prefix for the conference application:"), QLineEdit::Normal, prefix, &ok);
	if (ok && !text.isEmpty()) {
		prefix = text; 
		writeSettings();
		QString warning = "Current prefix for voice data is: " + prefix ;
		currentPrefLabel->setText(warning);
	}
}

void MainWindow::joinConference() {

	QTreeWidgetItem *current = pubConfList->currentItem();	
	if (!current) {
		critical("current is null");
	}

	QString confName = current->text(3);
	// should be indicated from current item in the future
	bool audio = true;
	bool video = false;
	if (video) {
		// no video yet; do nothing
	}
	if (audio) {

/*
		if (audioProcess != NULL && audioProcess->state() != QProcess::NotRunning) {
			int ret = QMessageBox::warning(this, tr("Conference Tool Set"),
											tr("The audio daemon is already running.\n" "Do you want to continue?"), QMessageBox::Cancel | QMessageBox::Yes, QMessageBox::Cancel);
			if (ret == QMessageBox::Cancel)
				return;
			else {
				audioProcess->kill();
				//if (audioProcess->state() != QProcess::NotRunning)
				//	debug("panic! can not kill audio tool process!");
				delete audioProcess;
				audioProcess = new QProcess(this);
			}
		}
		*/

		QString qsConfig = "<config><prefix>" + prefix + "</prefix><confName>"\
							+ confName + "</confName>";
		
		Announcement *a = itemToAnnouncement[current];
		if (a->getIsPrivate()) {
			qsConfig += "<private>true</private>";
			QByteArray confKey((const char *)a->conferenceKey, (int)sizeof(a->conferenceKey));	
			QString qsConfKey = QString(confKey.toBase64());
			qsConfig += "<confKey>" + qsConfKey + "</confKey>";
			QByteArray sessionKey((const char *)a->audioSessionKey, (int)sizeof(a->audioSessionKey));
			QString qsSessionKey = QString(sessionKey.toBase64());
			qsConfig += "<sessionKey>" + qsSessionKey + "</sessionKey>";
		}
		else {
			qsConfig += "<private>false</private>";
		}
		qsConfig += "</config>";

		QDomDocument doc;
		QDir actDir(QDir::homePath() + "/" + ".actd");
		if (!actDir.exists()) {
			QDir homedir = QDir::home();
			homedir.mkdir(".actd");
		}
		QString fileName = actDir.absolutePath() + "/" + ".config";
		QFile config(fileName);
		if (config.exists()) {
			config.remove();
		}

		config.open(QIODevice::WriteOnly | QIODevice::Truncate);
		QTextStream out (&config);

		out << qsConfig;
		// flush, maybe not needed, but sometimes release version of murmurd on iMac
		// uses old conference name
		config.flush();
		config.close();

		audioProcess = new QProcess(this);
		mumbleProcess = new QProcess(this);
		connect(mumbleProcess, SIGNAL(finished(int, QProcess::ExitStatus)), this, SLOT(mumbleCleanup(int, QProcess::ExitStatus)));
		connect(mumbleProcess, SIGNAL(started()), this, SLOT(mumbleStartup()));
		connect(audioProcess, SIGNAL(started()), this, SLOT(murmurStartup()));
#ifdef QT_NO_DEBUG
#ifdef __APPLE__
		audioPath = binaryPath + "/" + "ndn-murmurd";
		mumblePath = binaryPath + "/" + "Mumble";
#else
		audioPath = "ndn-murmurd";
		mumblePath = "ndn-mumble";
#endif
#else // QT_NO_DEBUG
#ifdef __APPLE__
		audioPath = binaryPath + "/../../../" + "ndn-murmurd";
		mumblePath = binaryPath + "/../../../Mumble.app/Contents/MacOS/" + "Mumble";
#else
		audioPath = binaryPath + "/ndn-murmurd";
		mumblePath = binaryPath + "/ndn-mumble";
#endif
#endif // QT_NO_DEBUG

		
		audioProcess->start(audioPath);
		mumbleProcess->start(mumblePath);
	}
}

void MainWindow::mumbleStartup() {
	QMessageBox::information(this, tr("Lauched Tools"), mumblePath );
}

void MainWindow::mumbleCleanup(int exitCode, QProcess::ExitStatus status) {
	if(audioProcess->state() != QProcess::NotRunning) {
		audioProcess->kill();
	}
	delete mumbleProcess;
	delete audioProcess;
}

void MainWindow::murmurStartup() {
	QMessageBox::information(this, tr("Lauched Tools"), audioPath );
}


void MainWindow::editConference() {
}


void MainWindow::dismissConference() {
	QTreeWidgetItem *current = pubConfList->currentItem();
	if (!current) {
		QString qs = "Dismiss: current is null!";
		critical(qs);
	}

	// remove from QList
	//QString confName = current->text(3);
	//QString organizer = current->text(4);
	Announcement *a = itemToAnnouncement[current];
	
	//sd->removeFromMyConferences(confName, organizer);
	sd->removeFromMyConferences(a);
	//delete a;
	itemToAnnouncement.remove(current);

	// remove from GUI
	delete current;
	
}

void MainWindow::newConference() {

	Announcement *a = new Announcement();
	ConfWizard wizard;
	wizard.exec();
	Announcement *announce = wizard.getAnnouncement();
	if (announce == NULL)
		return;
	a->copy(announce);
	addConferenceToList(a);
	sd->addToMyConferences(a);
}

void MainWindow::addConferenceToList(Announcement *announce) {

	if (announce == NULL) {
		QString qs ="add: conference announcement is null!";
		critical(qs);
	}

	QTreeWidgetItem *item = new QTreeWidgetItem;

	itemToAnnouncement.insert(item, announce);
	
	item->setText(3, announce->getConfName());
	QString oDesc = announce->getDesc();
	QString date = announce->getDate().toString("ddd MMM d"); 
	QString start = announce->getTime().toString("h:mm AP"); 
	int hours = announce->getHours();
	int minutes = announce->getMinutes();
	
	QString desc = QString("Date: %1\nTime: %2\nDuration: %3 Hours %4 Minutes\n-------------------------------\n%5").arg(date).arg(start).arg(hours).arg(minutes).arg(oDesc);

	item->setData(3, Qt::UserRole, desc);
	
	item->setText(4, announce->getOrganizer());
	item->setText(5, announce->getEmail());


	if (announce->getOwner() ) {
		QPixmap setting(":/images/setting.png");
		item->setIcon(0, setting);
	} else {
		if (announce->getIsPrivate()) {
			QPixmap priv(":/images/priv.png");
			item->setIcon(0, priv);
		}
	}

	if (announce->getAudio()) {
		QPixmap audio(":/images/audio.png");
		item->setIcon(1, audio);

	}

	if (announce->getVideo()) {
		QPixmap video(":/images/video.png");
		item->setIcon(2, video);
	}


	pubConfList->addTopLevelItem(item);
	if (!pubConfList->currentItem()) {
		pubConfList->setCurrentItem(pubConfList->topLevelItem(0));
		pubConfList->setEnabled(true);
		confDesc->setEnabled(true);
	}
}

void MainWindow::removeConferenceFromList(QString confName, QString organizer) {

	// remove from GUI
	QList<QTreeWidgetItem *> conferences = pubConfList->findItems(confName, Qt::MatchFixedString, 3);
	if (conferences.isEmpty()) {
		QString qs = QString("no conference named %1 found!!").arg(confName);
		critical(qs);
	}

	while (!conferences.isEmpty()) {
		QTreeWidgetItem *item = NULL;
		item = conferences.takeFirst();
		if (item != NULL && item->text(4) == organizer) {
			itemToAnnouncement.remove(item);
			delete item;
		}
	}

}

void MainWindow::listPrivateConferences() {
	if (listPrivate->isChecked()) {
		pubConfLabel->setText("Public And Private Conferences");
		sd->setListPrivate(true);
	}
	else {
		sd->setListPrivate(false);
		pubConfLabel->setText("Public Conferences");
	}
}
